### PR TITLE
[move] Shared objects must be newly created

### LINKED ISSUE #5835: [sui move] ban owned -> shared object promotions
We don't have a proof of correctness for this, and the need to handle it in Sui Core is raising subtle edge cases. We've agreed that the best course of action for now is to ban this.

Specifically: we need to add a dynamic check on `transfer::share_object(o)` that ensures `o.id` is an object ID created during the current transaction. The most natural approach to this is probably to internally track the set of created object id's and fail inside the native implementation of `share_object` if the condition is not met.

### PR DESCRIPTION
**Pull request by tnowacki**
- Throw abort on sui::transfer::shared if the object is not new
- This means it cannot be previously owned or wrapped
- Fixes #5835

- We might also need to update docs/examples?

### CODE REVIEW

File: crates/sui-framework/src/natives/object_runtime/mod.rs [hunk 1]
```diff
@@ -123,12 +129,25 @@ impl<'a> ObjectRuntime<'a> {
         ty: Type,
         tag: StructTag,
         obj: Value,
-    ) -> PartialVMResult<()> {
+    ) -> PartialVMResult<TransferResult> {
         let id: ObjectID = get_object_id(obj.copy_value()?)?
             .value_as::<AccountAddress>()?
             .into();
+        let transfer_result =
+            if self.state.new_ids.contains_key(&id) || id == SUI_SYSTEM_STATE_OBJECT_ID {
```
Comment (sblackshear) [inline crates/sui-framework/src/natives/object_runtime/mod.rs:137]: Second part of this conditional is kind of surprising--is this a special case for genesis? Might be worth mentioning if so.

**Review by sblackshear — APPROVED** [review]

**Comment by gdanezis** [comment]
Thanks for the simplification (tag: @lxfind )

File: crates/sui-framework/src/natives/object_runtime/mod.rs (hunk 1 above)
Comment (lxfind) [reply crates/sui-framework/src/natives/object_runtime/mod.rs:137]: Yeah worth a comment. The system state object is special in that its ID is not generated from TxContext, but hard coded at 0x5. It maybe clearer if we could find a cleaner way to augment new_ids to include it in the genesis case.

**Review by mystenmark — COMMENTED** [review]
Thanks for doing this!

**Comment by lxfind** [comment]
Just double check, it can detect sharing unwrapped objects as well, right?

**Comment by tnowacki** [comment]
> Just double check, it can detect sharing unwrapped objects as well, right?

Correct, I added a block of comments in there for that case too.

**Comment by tnowacki** [comment]
> > Just double check, it can detect sharing unwrapped objects as well, right?
> 
> Correct, I added a block of comments in there for that case too.

I updated the PR title and description to make this a bit more clear

**Pull request merged**